### PR TITLE
fix rollback panic bug

### PR DESCRIPTION
**Pull request by WIZARD-CXY**
fix the panic bug in freelist `rollback` function and add test.

### CODE REVIEW

**Comment by WIZARD-CXY** [comment]
ref https://github.com/etcd-io/bbolt/issues/152

**Comment by WIZARD-CXY** [comment]
Since I don't see the code testing the `Rollback` function when syncFreelist is true or false. I add both.

**Comment by WIZARD-CXY** [comment]
@kozlovic @xiang90

**Comment by WIZARD-CXY** [comment]
@yiju ptal

**Comment by xiang90** [comment]
Unfortunately, nosync option makes rollback very expensive due since it requires a full scan of the db. probably we should track the allocated pages for a pending tx. Then rollback only needs to put back the allocated pages.

I guess it is OK to accept the current approach as a start.

**Comment by xiang90** [comment]
@WIZARD-CXY Is there a less expensive way to rollback the tx? Can we simply put back the allocated pages for the write tx when rolling back?

**Comment by WIZARD-CXY** [comment]
@xiang90 yeah, I have the same concern. Now freelist `rollback` kinda what u want. But Tx rollback now call freelist `rollback` and `reload` two functions. I think `reload` which is heavy maybe  unnecessary.. but not for 100% sure. So I added it for now.

**Comment by WIZARD-CXY** [comment]
If we want keep the exact same behavior as the `sync freelist` version for the `nosyncfreelist`, we must fully load the whole freelist.

**Comment by WIZARD-CXY** [comment]
Actually, I tried to remove the `reload` function in `tx.Rollback` the boltdb test passed

**Comment by WIZARD-CXY** [comment]
@benbjohnson

**Comment by WIZARD-CXY** [comment]
@xiang90 @yudai @gyuho @hexfusion ptal

**Comment by WIZARD-CXY** [comment]
kindly ping @xiang90 @yudai @gyuho @hexfusion

File: tx.go [hunk 1]
```diff
@@ -264,7 +264,13 @@ func (tx *Tx) rollback() {
 	}
 	if tx.writable {
 		tx.db.freelist.rollback(tx.meta.txid)
-		tx.db.freelist.reload(tx.db.page(tx.db.meta().freelist))
+		if !tx.db.hasSyncedFreelist() {
+			// Reconstruct free page list by scanning the DB.
```
Comment (hormes) [inline tx.go]: add a TODO here, why we need reload DB here, as reload entire DB costs (although etcd is rarely rollback)
)

File: freelist.go [hunk 2]
```diff
@@ -349,6 +349,30 @@ func (f *freelist) reload(p *page) {
...
+	}
+
+	// Check each page in the freelist and build a new available freelist
+	// with any pages not in the pending lists.
+	var a []pgid
+	for _, id := range pgids {
+		if !pcache[id] {
+			a = append(a, id)
+		} else {
+			panic("should not happen")
```
Comment (hormes) [inline freelist.go]: why we panic here ?

File: tx.go (hunk 1 above)
Comment (WIZARD-CXY) [reply tx.go]: ok

**Comment by WIZARD-CXY** [comment]
@hormes changed ptal @xiang90

File: freelist.go [hunk 3]
```diff
@@ -349,6 +350,30 @@ func (f *freelist) reload(p *page) {
...
+	}
+
+	// Check each page in the freelist and build a new available freelist
+	// with any pages not in the pending lists.
+	var a []pgid
+	for _, id := range pgids {
+		if !pcache[id] {
+			a = append(a, id)
+		} else {
+			log.Printf("page id %d in pending list", id)
```
Comment (twpayne) [inline freelist.go]: Logging should generally be the responsibility of the application, not the library, so I suggest removing the `else` branch here.

Comment (WIZARD-CXY) [reply freelist.go]: I agree with u.

**Comment by WIZARD-CXY** [comment]
@xiang90 @hormes ptal

File: tx.go [hunk 4]
```diff
@@ -254,17 +254,36 @@ func (tx *Tx) Rollback() error {
 	if tx.db == nil {
 		return ErrTxClosed
 	}
-	tx.rollback()
+	tx.lightRollback()
 	return nil
 }
 
+// lightRollback is called when user calls Rollback directly, in this case we do not need to reload the free pages from disk.
+func (tx *Tx) lightRollback() {
```
Comment (xiang90) [inline tx.go]: nonPhysicalRollback

File: freelist.go [hunk 5]
```diff
@@ -349,6 +349,28 @@ func (f *freelist) reload(p *page) {
 	f.readIDs(a)
 }
 
+// noSyncReload reads the freelist from pgids and filters out pending items.
+func (f *freelist) noSyncReload(pgids []pgid) {
```
Comment (xiang90) [inline freelist.go]: freelist knows its own type. we can do the branch inside the reload func.

**Comment by xiang90** [comment]
just two styling issue.

File: freelist.go (hunk 5 above)
Comment (WIZARD-CXY) [reply freelist.go]: The type of freelist is hashmap or array. It is irrelevant with sync or not-sync.

**Comment by xiang90** [comment]
@WIZARD-CXY 

I did a release here https://github.com/etcd-io/bbolt/releases/tag/v1.3.3. can you update the dependency of bbolt in etcd to include the fix?

**Comment by WIZARD-CXY** [comment]
@xiang90 consider it done

**Pull request merged**